onnx/defs/tensor/old.cc

```diff
@@ -3274,13 +3279,20 @@ ONNX_OPERATOR_SET_SCHEMA(
             if (j < axes.size() && static_cast<size_t>(axes[j]) == i) {
               // There's a lot of potential behaviors. For now just
               // handle some simple cases.
-              if (ctx.getInputType(0)->tensor_type().shape().dim((int)i).has_dim_value() && starts[j] >= 0 &&
-                  ends[j] >= 0) {
-                auto newval =
-                    std::min((int64_t)ctx.getInputType(0)->tensor_type().shape().dim((int)i).dim_value(), ends[j]) -
-                    starts[j];
-                if (newval >= 0) {
-                  newdim->set_dim_value(newval);
+              const auto& dim = ctx.getInputType(0)->tensor_type().shape().dim((int)i);
+              if (dim.has_dim_value()) {
```

Comment:
We could improve this further (in the case where the dim's value is not known): the dim-value is needed only if one of starts/ends is positive and the other is negative. Otherwise, subtraction is fine.

Reply:
So, maybe something like below:
```cpp
   same_sign = ((starts[j] < 0) && (ends[j] < 0)) || ((starts[j] = 0) && (ends[j] >= 0))
   if (dim.has_dim_value()) {
      auto dim_value = dim.dim_value();
      if (starts[j] < 0) {
        starts[j] += dim_value;
      }
      if (ends[j] < 0) {
        ends[j] += dim_value;
      }
   }
   if (same_sign || dim.has_dim_value()) {
      auto newval = std::max(0, std::min(dim_value, ends[j]) - starts[j]);
      newdim->set_dim_value(newval);
   }
```

Reply:
> So, maybe something like below:
> 
> ```c++
>    same_sign = ((starts[j] < 0) && (ends[j] < 0)) || ((starts[j] >= 0) && (ends[j] >= 0))
>    if (dim.has_dim_value()) {
>       auto dim_value = dim.dim_value();
>       if (starts[j] < 0) {
>         starts[j] += dim_value;
>       }
>       if (ends[j] < 0) {
>         ends[j] += dim_value;
>       }
>    }
>    if (same_sign || dim.has_dim_value()) {
>       auto newval = std::max(0, std::min(dim_value, ends[j]) - starts[j]);
>       newdim->set_dim_value(newval);
>    }
> ```

but **auto newval = std::max(0, std::min(dim_value, ends[j]) - starts[j]);** still needs dim_value.

Reply:
Good point. So, this may be more complicated than I thought. I wanted to handle simple cases like slicing `X[-2: -1]` or `X[1:4]` ... which can almost be done without knowing the original dimension. But I guess we need the original dimension for special cases like when the index-value is out of range.